features/lorawan/LoRaWANStack.cpp

```diff
@@ -0,0 +1,2046 @@
+#define tr_error(...) (void(0)) //dummies if feature common pal is not added
+#define tr_warn(...) (void(0)) //dummies if feature common pal is not added
+#endif //defined(FEATURE_COMMON_PAL)
+
+#define INVALID_PORT        0xFF
+
+#ifdef MBED_CONF_LORA_PHY
+ #if MBED_CONF_LORA_PHY      == 0
+  #include "lorawan/lorastack/phy/LoRaPHYEU868.h"
+  static LoRaPHYEU868 lora_phy;
+ #elif MBED_CONF_LORA_PHY    == 1
+  #include "lorawan/lorastack/phy/LoRaPHYAS923.h"
+  static LoRaPHYAS923 lora_phy;
+ #elif MBED_CONF_LORA_PHY    == 2
+  #include "lorawan/lorastack/phy/LoRaPHYAU915.h"
+ static LoRaPHYAU915 lora_phy;
+ #elif MBED_CONF_LORA_PHY    == 3
+  #include "lorawan/lorastack/phy/LoRaPHYCN470.h"
+  static LoRaPHYCN470 lora_phy;
+ #elif MBED_CONF_LORA_PHY    == 4
+  #include "lorawan/lorastack/phy/LoRaPHYCN779.h"
+  static LoRaPHYCN779 lora_phy;
+ #elif MBED_CONF_LORA_PHY    == 5
+  #include "lorawan/lorastack/phy/LoRaPHYEU433.h"
+  static LoRaPHYEU433 lora_phy;
+ #elif MBED_CONF_LORA_PHY    == 6
+  #include "lorawan/lorastack/phy/LoRaPHYIN865.h"
+  static LoRaPHYIN865 lora_phy;
+ #elif MBED_CONF_LORA_PHY    == 7
+  #include "lorawan/lorastack/phy/LoRaPHYKR920.h"
+  static LoRaPHYKR920 lora_phy;
+ #elif MBED_CONF_LORA_PHY    == 8
+  #include "lorawan/lorastack/phy/LoRaPHYUS915.h"
+  static LoRaPHYUS915 lora_phy;
+ #elif MBED_CONF_LORA_PHY    == 9
+  #include "lorawan/lorastack/phy/LoRaPHYUS915Hybrid.h"
+  static LoRaPHYUS915Hybrid lora_phy;
+ #endif //MBED_CONF_LORA_PHY == VALUE
+#else
+ #error "Must set LoRa PHY layer parameters."
+#endif //MBED_CONF_LORA_PHY
+
+using namespace mbed;
+
+/**
+ * Helper function prototypes
+ */
+static Mcps_t interpret_mcps_confirm_type(const lora_mac_mcps_t& local);
+static Mib_t interpret_mib_req_confirm_type(const lora_mac_mib_t& mib_local);
+static lora_mac_event_info_status_t interpret_event_info_type(const LoRaMacEventInfoStatus_t& remote);
+
+#if MBED_CONF_LORA_PHY      == 0
+#include "lorawan/lorastack/mac/LoRaMacTest.h"
+
+//#define LORAWAN_COMPLIANCE_TEST_DUTYCYCLE_ON                          true
+
+#define USE_SEMTECH_DEFAULT_CHANNEL_LINEUP            1
+#if (USE_SEMTECH_DEFAULT_CHANNEL_LINEUP == 1)
+#define LC4                { 867100000, 0, { ( ( DR_5 << 4 ) | DR_0 ) }, 0 }
+#define LC5                { 867300000, 0, { ( ( DR_5 << 4 ) | DR_0 ) }, 0 }
+#define LC6                { 867500000, 0, { ( ( DR_5 << 4 ) | DR_0 ) }, 0 }
+#define LC7                { 867700000, 0, { ( ( DR_5 << 4 ) | DR_0 ) }, 0 }
+#define LC8                { 867900000, 0, { ( ( DR_5 << 4 ) | DR_0 ) }, 0 }
+#define LC9                { 868800000, 0, { ( ( DR_7 << 4 ) | DR_7 ) }, 2 }
+#define LC10               { 868300000, 0, { ( ( DR_6 << 4 ) | DR_6 ) }, 1 }
+#endif
+#endif
+
+/**
+ *
+ * User application data buffer size if compliance test is used
+ */
+#if (MBED_CONF_LORA_PHY == 0 || MBED_CONF_LORA_PHY == 4 || MBED_CONF_LORA_PHY == 6 || MBED_CONF_LORA_PHY == 7)
+#define LORAWAN_COMPLIANCE_TEST_DATA_SIZE                  16
```

Comment:
I see the phy defines from 0-9 above, the checks here seem to be missing 3,5. Are they not supported? Is this expected?

Reply:
This is for compliance testing only.  We know for sure only about EU868 MHz band. Chinese 470 MHz band is not up yet although Semtech have defined regional parameters for it. Same is the case for some other regions too. I am inclined to change this for only EU868 MHz band. But then I am not sure, how would we certify our stack for other regions. For more information https://github.com/Lora-net/LoRaMac-node check the table LoRaWAN certification results.